src/libraries/System.IO.Packaging/src/System/IO/Packaging/ZipPackagePartPiece.cs

```diff
@@ -0,0 +1,214 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.IO.Compression;
+
+namespace System.IO.Packaging
+{
+    /// <summary>
+    /// A part piece descriptor, made up of a ZipArchiveEntry and a PieceNameInfo.
+    /// </summary>
+    internal sealed class ZipPackagePartPiece : IComparable<ZipPackagePartPiece>
+    {
+        private const string PartPieceFileExtension = ".piece";
+        private const string PartPieceLastExtension = "].last";
+
+        #region Internal Methods
+        /// <summary>
+        /// Return true and create a ZipPackagePartPiece if the name in the input ZipArchiveEntry parses
+        /// as a piece name.
+        /// </summary>
+        /// <remarks>
+        /// No Uri validation is carried out at this level. All that is checked is valid piece syntax.
+        /// This means that the PrefixName returned as part of the ZipPackagePartPiece will not necessarily
+        /// be a part name. For example, it could be the name of the content type stream.
+        /// </remarks>
+        internal static bool TryParse(ZipArchiveEntry zipArchiveEntry, [NotNullWhen(true)] out ZipPackagePartPiece? partPiece)
+        {
+            if (zipArchiveEntry == null)
+                throw new ArgumentNullException(nameof(zipArchiveEntry));
+
+            partPiece = null;
+
+            bool success = TryParseName(zipArchiveEntry.FullName, out PackUriHelper.ValidatedPartUri? partUri, out string? prefixName, out int pieceNumber, out bool isLastPiece);
+
+            if (success)
+            {
+                partPiece = new ZipPackagePartPiece(zipArchiveEntry, partUri!, prefixName!, pieceNumber, isLastPiece);
+            }
+
+            return success;
+        }
+
+        /// <summary>
+        /// Return true and populate the output parameters if the path parses as a piece name.
+        /// </summary>
+        /// <remarks>
+        /// No Uri validation is carried out at this level. All that is checked is valid piece syntax.
+        /// This means that the output prefix name will not necessarily be a part name. For example,
+        /// it could be the name of the content type stream.
+        /// </remarks>
+        internal static bool TryParseName(string path, [NotNullWhen(true)] out PackUriHelper.ValidatedPartUri? partUri, [NotNullWhen(true)] out string? prefixName, out int pieceNumber, out bool isLastPiece)
+        {
+            bool success = true;
+            int searchPosition = path.Length;
+            // All piece names obey the syntax:
+            //  prefix_name "/" "[" 1*digit "]" [".last"] ".piece"
+            // Work backwards from the end of the full path, extracting and checking this metadata in stages.
+            // Stage 1: extract the file extension of ".piece".
+            int resultPosition = path.LastIndexOf(PartPieceFileExtension, StringComparison.OrdinalIgnoreCase);
```

Comment:
`PartPieceFileExtension` does not seem to be used anywhere else but here. I can directly be passed here.
```suggestion
            int resultPosition = path.LastIndexOf(".piece", StringComparison.OrdinalIgnoreCase);
```